roachtest: gopg failed

roachtest.gopg [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5954907?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5954907?buildTab=artifacts#/gopg) on master @ [7326bbbf94610c40ae2d2fdf42c119dfea93a170](https://github.com/cockroachdb/cockroach/commits/7326bbbf94610c40ae2d2fdf42c119dfea93a170):


```
test artifacts and logs in: /artifacts/gopg/run_1
	orm_helpers.go:191,gopg.go:170,gopg.go:182,test_runner.go:896: 
		Tests run on Cockroach v22.2.0-alpha.00000000-2143-g7326bbbf94
		Tests run against gopg v10.9.0
		193 Total Tests Run
		173 tests passed
		20 tests failed
		0 tests skipped
		8 tests ignored
		0 tests passed unexpectedly
		3 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: v10.TestColumnReuse - unknown (unexpected)
		--- FAIL: v10.TestAnonymousStructField - unknown (unexpected)
		--- FAIL: v10.TestEmptyQuery - unknown (unexpected)
		For a full summary look at the gopg artifacts 
		
		An updated blocklist (gopgBlockList22_2) is available in the artifacts' gopg log
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-experience
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*gopg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-18284

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -3247,6 +3247,9 @@ func (ex *connExecutor) notifyStatsRefresherOfNewTables(ctx context.Context) {
 // mutate descriptors prior to committing a SQL transaction.
 func (ex *connExecutor) runPreCommitStages(ctx context.Context) error {
 	scs := &ex.extraTxnState.schemaChangerState
+	if len(scs.state.Targets) == 0 {
+		return nil
+	}
 	deps := newSchemaChangerTxnRunDependencies(
 		ex.planner.SessionData(),
 		ex.planner.User(),

--- a/pkg/sql/tests/BUILD.bazel
+++ b/pkg/sql/tests/BUILD.bazel
@@ -29,6 +29,7 @@ go_test(
     srcs = [
         "autocommit_extended_protocol_test.go",
         "bank_test.go",
+        "empty_query_test.go",
         "enum_test.go",
         "hash_sharded_test.go",
         "impure_builtin_test.go",
@@ -109,6 +110,7 @@ go_test(
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_google_btree//:btree",
+        "@com_github_jackc_pgx_v4//:pgx",
         "@com_github_kr_pretty//:pretty",
         "@com_github_lib_pq//:pq",
         "@com_github_stretchr_testify//assert",

--- a/pkg/sql/tests/empty_query_test.go
+++ b/pkg/sql/tests/empty_query_test.go
@@ -1,0 +1,48 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"net/url"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/jackc/pgx/v4"
+	"github.com/stretchr/testify/require"
+)
+
+// TestEmptyQuery is a regression test to ensure that sending an empty
+// query to the database as the first query is safe.
+func TestEmptyQuery(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{Insecure: false, UseDatabase: "defaultdb"})
+	defer s.Stopper().Stop(context.Background())
+
+	pgURL, cleanupFunc := sqlutils.PGUrl(
+		t, s.ServingSQLAddr(), "testConnClose" /* prefix */, url.User(username.RootUser),
+	)
+	defer cleanupFunc()
+
+	ctx := context.Background()
+	conn, err := pgx.Connect(ctx, pgURL.String())
+	require.NoError(t, err)
+
+	err = conn.QueryRow(ctx, "").Scan()
+	require.Error(t, err)
+	require.Regexp(t, "no rows in result set", err)
+}

